app/webauthn/rest.py

```diff
@@ -0,0 +1,66 @@
+from flask import Blueprint, jsonify, request
+
+from app.dao.users_dao import get_user_by_id
+from app.dao.webauthn_credential_dao import (
+    dao_create_webauthn_credential,
+    dao_delete_webauthn_credential,
+    dao_get_webauthn_credential_by_id,
+    dao_update_webauthn_credential_name,
+)
+from app.errors import InvalidRequest, register_errors
+from app.schema_validation import validate
+from app.webauthn.webauthn_schema import (
+    post_create_webauthn_credential_schema,
+    post_update_webauthn_credential_schema,
+)
+
+webauthn_blueprint = Blueprint('webauthn', __name__, url_prefix='/user/<uuid:user_id>/webauthn')
+register_errors(webauthn_blueprint)
+
+
+@webauthn_blueprint.route('', methods=['GET'])
+def get_webauthn_credentials(user_id):
+    user = get_user_by_id(user_id)
+    return jsonify(data=[cred.serialize() for cred in user.webauthn_credentials]), 200
+
+
+@webauthn_blueprint.route('', methods=['POST'])
+def create_webauthn_credential(user_id):
+    data = request.get_json()
+    validate(data, post_create_webauthn_credential_schema)
+    webauthn_credential = dao_create_webauthn_credential(
+        user_id=user_id,
+        name=data['name'],
+        credential_data=data['credential_data'],
+        registration_response=data['registration_response']
+    )
+
+    return jsonify(data=webauthn_credential.serialize()), 201
+
+
+@webauthn_blueprint.route('/<uuid:webauthn_credential_id>', methods=['POST'])
+def update_webauthn_credential(user_id, webauthn_credential_id):
+    data = request.get_json()
+    validate(data, post_update_webauthn_credential_schema)
+
+    webauthn_credential = dao_get_webauthn_credential_by_id(webauthn_credential_id)
+
+    dao_update_webauthn_credential_name(webauthn_credential, data['name'])
+
+    return jsonify(data=webauthn_credential.serialize()), 200
+
+
+@webauthn_blueprint.route('/<uuid:webauthn_credential_id>', methods=['DELETE'])
+def delete_webauthn_credential(user_id, webauthn_credential_id):
+    webauthn_credential = dao_get_webauthn_credential_by_id(webauthn_credential_id)
+    user = get_user_by_id(user_id)
+
+    if webauthn_credential.user_id != user.id:
+        raise InvalidRequest('Webauthn credential does not belong to this user', status_code=400)
+
+    if len(user.webauthn_credentials) == 1:
```

Comment:
Minor: I don't think we need this check in the API. Reasons are:

- We're going to check it in Admin anyway.
- It's not a technical issue in the API if there are no credentials for a user.
- We're hoping to remove this constraint in future.

Reply:
I think it's usually good to have two lines of defence, just in case, at least that's what I've learnt both on my course and so far at GDS.

Why are we hoping to remove this constraint in the future?

Reply:
I'd usually agree, if the code is to protect us from a technical (or security) issue.

We're not sure if we're going to keep this restriction, as there are other ways to prevent a user being locked out if they remove their last key. In particular, we want to allow a user to switch of WebAuthn somehow.

So I don't think we should make our code to "hard", as this just means it's more work to change later on. On the other hand, it's not _that_ much work to remove the restriction here, which is why I put it down as a minor issue.

In future, we could scope this check to only error if the user has the `webauth_auth` type. Otherwise, it's absolutely fine for them to remove their last credential, since they're not using them for anything.

Reply:
lets leave it as is for now, i've put a TODO comment in to add webauthn_auth type checking 👍